### ISSUE TITLE
LSTM text generation, increasing loss function

### ISSUE THREAD

**Issue by mostafa-mahmoud** [fchollet/keras#7375]
I was running the Nietzsche's LSTM text generator [script](https://github.com/fchollet/keras/blob/master/examples/lstm_text_generation.py) with slight modifications, and I the loss function showed strange behaviour after 70 iterations (it increased significantly), Is there a reason for this behaviour? or is it a bug?

[Here](https://gist.github.com/mostafa-mahmoud/b7058bb8e5b2079ad1cb45d0873de67d) is the code I used, and here is the loss output: (UPDATE: results of more iterations)

iteration 1 200285/200285 [==============================] - 433s - loss: 2.0411     
iteration 2 200285/200285 [==============================] - 433s - loss: 1.6363     
iteration 3 200285/200285 [==============================] - 441s - loss: 1.5480     
iteration 4 200285/200285 [==============================] - 457s - loss: 1.5036     
iteration 5 200285/200285 [==============================] - 437s - loss: 1.4742     
iteration 6 200285/200285 [==============================] - 484s - loss: 1.4563     
iteration 7 200285/200285 [==============================] - 459s - loss: 1.4406     
iteration 8 200285/200285 [==============================] - 418s - loss: 1.4286     
iteration 9 200285/200285 [==============================] - 412s - loss: 1.4175     
iteration 10 200285/200285 [==============================] - 406s - loss: 1.4061     
iteration 11 200285/200285 [==============================] - 440s - loss: 1.3981     
iteration 12 200285/200285 [==============================] - 454s - loss: 1.3910     
iteration 13 200285/200285 [==============================] - 437s - loss: 1.3848     
iteration 14 200285/200285 [==============================] - 430s - loss: 1.3789    
iteration 15 200285/200285 [==============================] - 482s - loss: 1.3745     
iteration 16 200285/200285 [==============================] - 472s - loss: 1.3688     
iteration 17 200285/200285 [==============================] - 471s - loss: 1.3643     
iteration 18 200285/200285 [==============================] - 453s - loss: 1.3595     
iteration 19 200285/200285 [==============================] - 423s - loss: 1.3565     
iteration 20 200285/200285 [==============================] - 344s - loss: 1.3532     
iteration 21 200285/200285 [==============================] - 343s - loss: 1.3495     
iteration 22 200285/200285 [==============================] - 338s - loss: 1.3473     
iteration 23 200285/200285 [==============================] - 338s - loss: 1.3442     
iteration 24 200285/200285 [==============================] - 337s - loss: 1.3410     
iteration 25 200285/200285 [==============================] - 335s - loss: 1.3396     
iteration 26 200285/200285 [==============================] - 333s - loss: 1.3379     
iteration 27 200285/200285 [==============================] - 336s - loss: 1.3346     
iteration 28 200285/200285 [==============================] - 336s - loss: 1.3338     
iteration 29 200285/200285 [==============================] - 333s - loss: 1.3307     
iteration 30 200285/200285 [==============================] - 333s - loss: 1.3300     
iteration 31 200285/200285 [==============================] - 333s - loss: 1.3293     
iteration 32 200285/200285 [==============================] - 333s - loss: 1.3279     
iteration 33 200285/200285 [==============================] - 340s - loss: 1.3273     
iteration 34 200285/200285 [==============================] - 334s - loss: 1.3273     
iteration 35 200285/200285 [==============================] - 380s - loss: 1.3270     
iteration 36 200285/200285 [==============================] - 419s - loss: 1.3246     
iteration 37 200285/200285 [==============================] - 396s - loss: 1.3248     
iteration 38 200285/200285 [==============================] - 368s - loss: 1.3238     
iteration 39 200285/200285 [==============================] - 349s - loss: 1.3222     
iteration 40 200285/200285 [==============================] - 490s - loss: 1.3215     
iteration 41 200285/200285 [==============================] - 477s - loss: 1.3205     
iteration 42 200285/200285 [==============================] - 474s - loss: 1.3225     
iteration 43 200285/200285 [==============================] - 513s - loss: 1.3311     
iteration 44 200285/200285 [==============================] - 522s - loss: 1.3389     
iteration 45 200285/200285 [==============================] - 502s - loss: 1.3728     
iteration 46 200285/200285 [==============================] - 484s - loss: 1.3313     
iteration 47 200285/200285 [==============================] - 494s - loss: 1.3233     
iteration 48 200285/200285 [==============================] - 550s - loss: 1.3226     
iteration 49 200285/200285 [==============================] - 664s - loss: 1.3202     
iteration 50 200285/200285 [==============================] - 490s - loss: 1.3228     
iteration 51 200285/200285 [==============================] - 404s - loss: 1.3187     
iteration 52 200285/200285 [==============================] - 349s - loss: 1.3200     
iteration 53 200285/200285 [==============================] - 347s - loss: 1.3195     
iteration 54 200285/200285 [==============================] - 402s - loss: 1.3245     
iteration 55 200285/200285 [==============================] - 413s - loss: 1.3203     
iteration 56 200285/200285 [==============================] - 393s - loss: 1.3164     
iteration 57 200285/200285 [==============================] - 370s - loss: 1.3190     
iteration 58 200285/200285 [==============================] - 442s - loss: 1.3536     
iteration 59 200285/200285 [==============================] - 408s - loss: 1.3241     
iteration 60 200285/200285 [==============================] - 356s - loss: 1.3184     
iteration 61 200285/200285 [==============================] - 388s - loss: 1.3224     
iteration 62 200285/200285 [==============================] - 395s - loss: 1.3196     
iteration 63 200285/200285 [==============================] - 415s - loss: 1.3429     
iteration 64 200285/200285 [==============================] - 413s - loss: 1.6466     
iteration 65 200285/200285 [==============================] - 401s - loss: 1.3339     
iteration 66 200285/200285 [==============================] - 443s - loss: 1.3263     
iteration 67 200285/200285 [==============================] - 430s - loss: 1.3607     
iteration 68 200285/200285 [==============================] - 393s - loss: 1.3696     
iteration 69 200285/200285 [==============================] - 460s - loss: 1.3275     
iteration 70 200285/200285 [==============================] - 443s - loss: 1.6963     
iteration 71 200285/200285 [==============================] - 449s - loss: 4.9700     
iteration 72 200285/200285 [==============================] - 473s - loss: 7.8267     
iteration 73 200285/200285 [==============================] - 441s - loss: 9.1923     
iteration 74 200285/200285 [==============================] - 422s - loss: 8.6638      
iteration 75 200285/200285 [==============================] - 423s - loss: 8.1039      
iteration 76 200285/200285 [==============================] - 403s - loss: 7.8252      
iteration 77 200285/200285 [==============================] - 416s - loss: 6.7547     
iteration 78 200285/200285 [==============================] - 436s - loss: 9.0915     
iteration 79 200285/200285 [==============================] - 450s - loss: 9.4587      
iteration 80 200285/200285 [==============================] - 498s - loss: 11.3355     
iteration 81 200285/200285 [==============================] - 453s - loss: 10.5768     
iteration 82 200285/200285 [==============================] - 471s - loss: 9.1803     
iteration 83 200285/200285 [==============================] - 454s - loss: 9.2897     
iteration 84 200285/200285 [==============================] - 439s - loss: 10.3444     
iteration 85 200285/200285 [==============================] - 460s - loss: 10.4792     
iteration 86 200285/200285 [==============================] - 454s - loss: 9.9937      
iteration 87 200285/200285 [==============================] - 470s - loss: 10.5900     
iteration 88 200285/200285 [==============================] - 459s - loss: 10.3293     
iteration 89 200285/200285 [==============================] - 470s - loss: 9.2961      
iteration 90 200285/200285 [==============================] - 457s - loss: 8.7879     
iteration 91 200285/200285 [==============================] - 475s - loss: 8.7853     
iteration 92 200285/200285 [==============================] - 471s - loss: 8.8662     
iteration 93 200285/200285 [==============================] - 456s - loss: 9.1475     
iteration 94 200285/200285 [==============================] - 453s - loss: 9.1911     
iteration 95 200285/200285 [==============================] - 453s - loss: 9.1089     
iteration 96 200285/200285 [==============================] - 412s - loss: 8.6581     
iteration 97 200285/200285 [==============================] - 440s - loss: 8.2222     
iteration 98 200285/200285 [==============================] - 427s - loss: 7.8420     
iteration 99 200285/200285 [==============================] - 456s - loss: 8.0445     
iteration 100 200285/200285 [==============================] - 475s - loss: 8.2455     
iteration 101 200285/200285 [==============================] - 408s - loss: 8.2646     
iteration 102 200285/200285 [==============================] - 480s - loss: 8.4509     
iteration 103 200285/200285 [==============================] - 487s - loss: 8.7179     
iteration 104 200285/200285 [==============================] - 517s - loss: 8.4393     
iteration 105 200285/200285 [==============================] - 578s - loss: 8.5690     
iteration 106 200285/200285 [==============================] - 510s - loss: 8.5140     
iteration 107 200285/200285 [==============================] - 493s - loss: 8.4314     
iteration 108 200285/200285 [==============================] - 481s - loss: 8.5455     
iteration 109 200285/200285 [==============================] - 436s - loss: 8.6509   
iteration 110 200285/200285 [==============================] - 568s - loss: 8.5100     
iteration 111 200285/200285 [==============================] - 410s - loss: 8.5536     
iteration 112 200285/200285 [==============================] - 430s - loss: 8.6331     
iteration 113 200285/200285 [==============================] - 330s - loss: 8.5045     
iteration 114 200285/200285 [==============================] - 331s - loss: 8.4348     
iteration 115 200285/200285 [==============================] - 329s - loss: 8.2595     
iteration 116 200285/200285 [==============================] - 328s - loss: 8.2419     
iteration 117 200285/200285 [==============================] - 326s - loss: 8.1802     
iteration 118 200285/200285 [==============================] - 326s - loss: 8.0493     
iteration 119 200285/200285 [==============================] - 329s - loss: 8.0187     
iteration 120 200285/200285 [==============================] - 324s - loss: 8.0077     
iteration 121 200285/200285 [==============================] - 325s - loss: 7.7659     
iteration 122 200285/200285 [==============================] - 328s - loss: 7.7752     
iteration 123 200285/200285 [==============================] - 324s - loss: 7.7003     
iteration 124 200285/200285 [==============================] - 325s - loss: 7.8265     
iteration 125 200285/200285 [==============================] - 322s - loss: nan        
iteration 126 200285/200285 [==============================] - 324s - loss: nan     
iteration 127 200285/200285 [==============================] - 323s - loss: nan     
iteration 128 200285/200285 [==============================] - 320s - loss: nan     
iteration 129 200285/200285 [==============================] - 321s - loss: nan     
iteration 130 200285/200285 [==============================] - 320s - loss: nan     
iteration 131 200285/200285 [==============================] - 320s - loss: nan     
iteration 132 200285/200285 [==============================] - 322s - loss: nan     
iteration 133 200285/200285 [==============================] - 324s - loss: nan     
iteration 134 200285/200285 [==============================] - 322s - loss: nan     
iteration 135 200285/200285 [==============================] - 322s - loss: nan     
iteration 136 200285/200285 [==============================] - 327s - loss: nan     
iteration 137 200285/200285 [==============================] - 322s - loss: nan     
iteration 138 200285/200285 [==============================] - 322s - loss: nan     
iteration 139 200285/200285 [==============================] - 322s - loss: nan     
iteration 140 200285/200285 [==============================] - 321s - loss: nan     
iteration 141 200285/200285 [==============================] - 355s - loss: nan     
iteration 142 200285/200285 [==============================] - 418s - loss: nan     
iteration 143 200285/200285 [==============================] - 400s - loss: nan     
iteration 144 200285/200285 [==============================] - 422s - loss: nan     
iteration 145 200285/200285 [==============================] - 427s - loss: nan     
iteration 146 200285/200285 [==============================] - 414s - loss: nan     
iteration 147 200285/200285 [==============================] - 435s - loss: nan     
iteration 148 200285/200285 [==============================] - 435s - loss: nan     
iteration 149 200285/200285 [==============================] - 430s - loss: nan     
iteration 150 200285/200285 [==============================] - 426s - loss: nan

**Comment by mostafa-mahmoud** [comment]
@fchollet

**Comment by myurasov** [comment]
@mostafa-mahmoud catastrophic forgetting?

**Comment by mostafa-mahmoud** [comment]
@myurasov  seems so.

This is a sample of the generated text:

```
Iteration 69

----- diversity: 0.2
----- Generating with seed: " reprisal and
exchange upon the basis of"
 reprisal and
exchange upon the basis of the soul the sense of the conception and the most philosophy in the sentiment and false the sentiments of the world, and the conceality and the subtlerly and the sentiment and state of the proportion that it is a surfect of the sense of the conscience and philosophy in the philosophy in the sentiment and the power and philosophy of the power and the conception the whole their instinct and conceal

----- diversity: 0.5
----- Generating with seed: " reprisal and
exchange upon the basis of"
 reprisal and
exchange upon the basis of an aneistation, and morality itself in the soul has no longer are calls the and interpretuation is former that the soul and action and superstitive instinct of the ward has the sent in speaks and the sensetion with the dangerous all themselves that a power and new things of all the desires in power and religious and are at all a new permonside and independently have been the conception as the sen

----- diversity: 1.0
----- Generating with seed: " reprisal and
exchange upon the basis of"
 reprisal and
exchange upon the basis often as known and insersed even us then the most let most firlmand canclicalares over, god; bother--he
acconderous
imagination. all that mudeful; there us long power it--i ambegied phe'sad otnes inspeach? and something,
by the capacity: therefore
lifern. that 
sent for men. whother belong to oodsing. it may
be something problem of itself that has sames their pluaces and de-feminins a 
neither ly-ou

----- diversity: 1.2
----- Generating with seed: " reprisal and
exchange upon the basis of"
 reprisal and
exchange upon the basis offen which
baling! "has sociations
for commuct sphe one, he is littatfuls: and bould onuthes
conccroble even asimon image a deceived thus changed know an and to
despect, man can only such does not consisted in attaining, view which sympathy, above the wich
youo
considered proiteming. and treward fof explaition.
ykxkernuing ordinary, longinessidders things or, it is; in worphtaly-phraborehs itself. 

--------------------------------------------------
Iteration 71

----- diversity: 0.2
----- Generating with seed: "o attained, the heights, depths, and dis"
o attained, the heights, depths, and diséép0aæ0eé00es the corll xhej=actin to fact orselea
centure(se offlher thetéé iélé tht céless of tmy thregul andothe p(ihp0 in it insévect to camnn the corucetin th doup thb aregéé in ,éged the cordeie me the conse0lysed téles0,é in ,ht inst of tnekeque the cofhoffé
ré]0s o00bcie to the
chpach inst andohhhéxlæst inévelly is an éce the cordeilooäé --in iheve a noäirlacter the corfhitaf instotal(zing

----- diversity: 0.5
----- Generating with seed: "o attained, the heights, depths, and dis"
o attained, the heights, depths, and diséäl0eä0bé00s advangt prese(pleed inst wared to fæémen andotä asés to have tret lasile z(notihä instaocely, wither thbr(gouné,æst tht coftoty "tals the coneexë ëp. up and a insthet buttnerst andorego i0pgen 0buus ordandoporrenthist our the choæly ane(quw, bgoen the dange the "yo anne   xisi(fac: ho our tea; afæéven to hægins æreé in , a sityre seehe  at a    quiy is a0quactiness dows, thm the conto

----- diversity: 1.0
----- Generating with seed: "o attained, the heights, depths, and dis"
o attained, the heights, depths, and diséæsä é0léé fore of and,noohoëe"g oldolapt: he re0vouso, in as a tzmaochc anan"or owfarén0éræ"k,äsy célinéé-bel =q0dq00000p. the ethithust and twa,ë. he le mo noncr0l taken is to ral andeets ye
highe pro7eé fæst hoe an"d as thetthth,æv." tevinces mis indeæsculaucts himplin"oe, en. the
chou, dom, afters0, sreasvapary ess reanderoured a p(iny.h æst initethexiyfulal comany ærkenä andemncide(-witifain 

----- diversity: 1.2
----- Generating with seed: "o attained, the heights, depths, and dis"
o attained, the heights, depths, and diséæaë,é0aæ00vficer thst to s;manlave tonar reanghoodehand itt ist ths posio seeneaéven, gran lang épleeiselay-0vorcescing the pxhxiecrx! i mafh the haä-élicaläly, grean, awe
in ,u reum(fdanger tht prelié. gemiviés thmp inst of thzeeé" af thleeé "prese initerlitr seeen the0gætuä routh undivisticms,é-cident thik are anbe reærlongowsed th éc0s0e éary
oppoflumatct mist war as ow l  hop? insaks insartin

```

**Comment by gewoonrik** [comment]
could you try another optimizer, for example Adam, to check if RMSProp is the problem?

**Comment by mostafa-mahmoud** [comment]
I tried Adam optimizer, and same error appears still.

**Comment by Suzanaa** [comment]
Has this issue been addressed? I have the same problem (different, bigger dataset)

**Comment by mostafa-mahmoud** [comment]
@Suzanaa I eventually used the same model on other datasets with few modifications and it worked. 
I think adding the dropout and using lower learning rate are probably the reason for solving the issue, since it seemed to be an exploding gradient issue. Another way that I found, in Andrew's NG course 5, that is usually used in RNNs, is clipping the gradient (by putting an upper bound on the gradient values) to avoid the exploding gradients, however I didn't need to do that in the implementation below.
` 
model = Sequential()
model.add(LSTM(128, input_shape=(context_length, alphabet_size)))
model.add(Dropout(0.2))
model.add(Dense(alphabet_size))
model.add(Activation('softmax'))
model.compile(loss='categorical_crossentropy', optimizer=Adam(lr=0.0005), metrics=['accuracy'])
`

I would also advice to have a look on [https://github.com/ekzhang/char-rnn-keras/blob/master/model.py](this) , which has a multi-layered TimeDistributed implementation of the model.
